File: python/src/mpeg_o/protocols/streamable.py
"""``Streamable`` — sequential access with explicit positioning."""
from __future__ import annotations

from typing import Any, Protocol, runtime_checkable


@runtime_checkable
class Streamable(Protocol):
    """Capability for sequential access with explicit positioning.

    Enables efficient iteration over large datasets without
    materializing the entire collection in memory.

    Methods
    -------
    next_object()
        Return the next element and advance the cursor.
    has_more()
        Return ``True`` if ``next_object`` can be called.
    current_position()
        Return the 0-based position of the next element to be yielded.
    seek_to_position(position)
        Reposition the cursor. Returns ``True`` on success.
    reset()
        Reposition the cursor to 0.

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: ``MPGOStreamable`` ·
    Java: ``com.dtwthalion.mpgo.protocols.Streamable``
    """

    def next_object(self) -> Any: ...
    def has_more(self) -> bool: ...
    def current_position(self) -> int: ...
    def seek_to_position(self, position: int) -> bool: ...
    def reset(self) -> None: ...

File: java/src/main/java/com/dtwthalion/mpgo/protocols/Streamable.java
/*
 * MPEG-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package com.dtwthalion.mpgo.protocols;

/**
 * Objects implementing {@code Streamable} support sequential access
 * with explicit positioning. This enables efficient iteration over
 * large datasets without materializing the entire collection in
 * memory.
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code MPGOStreamable}, Python
 * {@code mpeg_o.protocols.Streamable}.</p>
 *
 * @param <T> element type
 * @since 0.6
 */
public interface Streamable<T> {

    /** @return the next element and advance the cursor. */
    T nextObject();

    /** @return {@code true} if {@link #nextObject} may be called. */
    boolean hasMore();

    /** @return 0-based position of the next element to be yielded. */
    int currentPosition();

    /**
     * Reposition the cursor.
     * @return {@code true} on success, {@code false} if out of range.
     */
    boolean seekToPosition(int position);

    /** Reposition the cursor to 0. */
    void reset();
}
